Security_Engine/Compute/CameraFieldOfView.cs

```diff
@@ -0,0 +1,351 @@
+        [Input("obstacles", "Polyline objects that represents obstacles for camera vision (walls, columns, stairs).")]
+        [Output("fieldOfView", "PolyCurve object that represents security camera field of view.")]
+        public static PolyCurve CameraFieldOfView(this CameraDevice cameraDevice, List<Polyline> obstacles, double distanceTolerance = BH.oM.Geometry.Tolerance.Distance, double angleTolerance = 0.01)
+        {
+            PolyCurve cameraCone = cameraDevice.ViewCone();
+            Polyline cameraConePolyline = cameraCone.CollapseToPolyline(angleTolerance);
+            Point cameraLocation = cameraDevice.EyePosition;
+            Point targetLocation = cameraDevice.TargetPosition;
+            double radius = targetLocation.Distance(cameraLocation);
+            Plane cameraPlane = BH.Engine.Geometry.Create.Plane(cameraLocation, Vector.ZAxis);
+
+            //cone points
+            List<Point> conePoints = cameraConePolyline.ControlPoints();
+
+            //add cone boundary points as intersectPoints
+            List<Point> intersectPoints = new List<Point>();
+            intersectPoints.Add(conePoints[1]);
+            intersectPoints.Add(conePoints[conePoints.Count - 2]);
+
+            //simplify obstacles
+            List<Polyline> projObstacles = obstacles.Select(x => x.FixAndProject(cameraPlane, distanceTolerance)).ToList();
+
+            //points that intersect with obstacles
+            List<Polyline> intersectObstacles = new List<Polyline>();
+            foreach (Polyline obstacle in projObstacles)
+            {
+                if (obstacle.IsContaining(new List<Point> { cameraLocation }))
+                {
+                    Base.Compute.RecordWarning("Camera Device is inside obstacle. Null value will be returned.");
+                    return null;
+                }
+                List<Polyline> intersection = obstacle.BooleanIntersection(cameraConePolyline);
+                if (intersection.Count != 0)
+                {
+                    intersectObstacles.Add(obstacle);
+                    foreach (Polyline intersectPolyline in intersection)
+                    {
+                        List<Point> points = intersectPolyline.ControlPoints;
+                        intersectPoints.AddRange(points);
+                    }
+                }
+            }
+            intersectPoints = intersectPoints.CullDuplicates();
+
+            //create and sort extended ray lines passing through intersected points
+            List<Line> rayLines = new List<Line>();
+            foreach (Point point in intersectPoints)
+            {
+                Line line = BH.Engine.Geometry.Create.Line(cameraLocation, point);
+                Line rayLine = line.Extend(0, radius - line.Length());
+                rayLines.Add(rayLine);
+            }
+            rayLines = rayLines.CullDuplicateLines();
+            rayLines = rayLines.OrderBy(x => x.SingedAngle(rayLines[0], Vector.ZAxis)).ToList();
```

Comment:
This could surely be optimised e.g. by creating already extended lines, culling duplicate endpoints only instead of the heavier duplicate lines, not sure if ordering is needed altogether (and if it is, whether could not be done cheaper)

Reply:
Moved to issue #2890.